README.md
<h1 align="center">
    <b>AEA BabyAGI</b>
</h1>

### Adaptation of [babyagi](https://github.com/yoheinakajima/babyagi) in the [Open AEA](https://github.com/valory-xyz/open-aea) framework.

## Project
```ml
├─ actions — "contains all the functions used to compose the low-high level actions of each agent"
├─ simple_babyagi — "Adaptation of the babyAGI agent loop/functions using only the OpenAI API without extra tooling"
├─ agent_babyagi - "Use Open AEA's Agent class to extend simple_babyagi into an Open AEA "Agent" with Finite State Machine Behaviour"
├─ aea_babyagi - "Inherit from Open AEA's "AEA" class to extend babyagi's functionality within agent_agi into an autonomous economic agent."
```

## Getting Started

Create a `.env` file from the `.env.example` provided in this repo:
```bash
cp .env.example .env
```

Set your OpenAI API key and, optionally, other environment variables you want to use in the `.env` file:
```bash
OPENAI_API_KEY="YOUR_API_KEY"
PINECONE_API_KEY="YOUR_API_KEY"
```

Install project dependencies (you can find install instructions for Poetry [here](https://python-poetry.org/docs/)):
```bash
poetry shell
poetry install
```

Import AEA packages:
```bash
svn export https://github.com/valory-xyz/open-aea/tags/v1.33.0/packages packages
```

Source the environment variables:
``` bash
source .env
```

Run the agents:
```bash
poetry run python simple_babyagi.py "develop a task list" "solve world hunger"
``` 
```bash
poetry run python agent_babyagi.py "develop a task list" "solve world hunger"
``` 
```bash
poetry run python aea_babyagi.py "develop a task list" "solve world hunger"
```


HISTORY.md
## 0.1.0 (2023-05-11)

- The first release of aea-babyagi

agent.py
from collections import deque
import os
import time
from threading import Thread
from typing import List
import openai
from dotenv import load_dotenv
import sys

load_dotenv()

# AEA dependencies
from aea.agent import Agent
from aea.configurations.base import SkillConfig
from aea.connections.base import Connection
from aea.identity.base import Identity
from aea.mail.base import Envelope
from aea.crypto.registries import make_crypto
from aea.skills.base import Skill, SkillContext
from aea.skills.behaviours import FSMBehaviour, State
from aea.context.base import AgentContext

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")

# Set up OpenAI API key
openai.api_key = OPENAI_API_KEY

# import functions used to build the agent's actions
from simple_agent import (
    STOP_PROCEDURE,
    task_creation_prompt_builder,
    task_creation_handler,
    task_prioritization_prompt_builder,
    task_prioritization_handler,
    task_execution_prompt_builder,
    task_execution_handler,
    task_stop_or_not_prompt_builder,
    task_stop_or_not_handler,
)

# action types definition, each action type makes two function calls: builder & handler
# the initial action type is execution of the first task
initial = "task_execution_1"
action_types = {
    "task_creation": {
        "prompt_builder": task_creation_prompt_builder,
        "handler": task_creation_handler,
    },
    "task_prioritization": {
        "prompt_builder": task_prioritization_prompt_builder,
        "handler": task_prioritization_handler,
    },
    "task_execution_1": {
        "prompt_builder": task_execution_prompt_builder,
        "handler": task_execution_handler,
    },
    "task_execution_2": {
        "prompt_builder": task_execution_prompt_builder,
        "handler": task_execution_handler,
    },
    "task_stop_or_not": {
        "prompt_builder": task_stop_or_not_prompt_builder,
        "handler": task_stop_or_not_handler,
    },
}
# State Machine Definition
# End state adds the option to stop the procedure and execute task_stop_or_not
if STOP_PROCEDURE:
    transitions = {
        "task_execution_1": {"done": "task_creation"},
        "task_creation": {"done": "task_execution_2"},
        "task_execution_2": {"done": "task_prioritization"},
        "task_prioritization": {"done": "task_stop_or_not"},
        "task_stop_or_not": {"done": "task_execution", "stop": None},
    }
# runtime state transitions of the agent loop (execution, creation, execution, prioritization)
else:
    transitions = {
        "task_execution_1": {"done": "task_creation"},
        "task_creation": {"done": "task_execution_2"},
        "task_execution_2": {"done": "task_prioritization"},
        "task_prioritization": {"done": "task_execution_1"},
    }


class SimpleStateBehaviour(State):

    def act(self) -> None:
        """
        Act implementation.
        """
        # get the action type
        action_type = action_types[self.name]
        # get the prompt builder for the action type
        builder_ = action_type["prompt_builder"]
        # build the prompt using the shared state from the Agent's context
        prompt = builder_(self.context.shared_state)
        # use the prompt above to input into GPT to get the response
        response = self.openai_call(prompt)
        # get the handler for the action type
        handler_ = action_type["handler"]
        # get the event to trigger from the handler
        event_to_trigger = handler_(response, self.context.shared_state)
        self.executed = True
        self._event = event_to_trigger

    @staticmethod
    def openai_call(
        prompt: str,
        use_gpt4: bool = False,
        temperature: float = 0.5,
        max_tokens: int = 200,
    ):
        if use_gpt4:
            messages = [{"role": "user", "content": prompt}]
            response = openai.ChatCompletion.create(
                model="gpt-4",
                messages=messages,
                temperature=temperature,
                max_tokens=max_tokens,
                n=1,
                stop=None,
            )
            return response.choices[0].message.content.strip()
        response = openai.Completion.create(
            engine="text-davinci-003",
            prompt=prompt,
            temperature=temperature,
            max_tokens=max_tokens,
            top_p=1,
            frequency_penalty=0,
            presence_penalty=0,
        )
        return response.choices[0].text.strip()

    def is_done(self) -> bool:
        """Get is done."""
        return self._event is not None
    

# make dummy concrete FSMBehaviour class for use in constructing the agent's FSM transitions/logic
class MyFSMBehaviour(FSMBehaviour):
    def setup(self):
        # dummy setup method
        pass

    def teardown(self):
        # dummy teardown method
        pass

# create the agent's shared state and return it, "memory" (dictionary that stores the agent's state)
# takes in string arguments from the command line to set the first task and the objective of the agent
def create_memory(first_task: str, objective: str,) -> dict:
    """Create the shared memory."""
    memory = {
            "objective": objective,
            "task_list": deque([]),
            "current_task": {},
            "result": {"data": first_task},
            "keep_going": True,
        }
    memory["task_list"].append({"id": 1, "name": first_task})
    return memory


# build the FSM object and return it, takes in the agent's shared state as an argument
# right now this function returns a tuple of the FSM object and the agent's skill object
# this can be split up as the Skill is needed for the aea but not the FSMBehaviour used in agent
def build_fsm_and_skill(memory: dict) -> tuple[MyFSMBehaviour, Skill]:
    """Build the FSM object."""
    config = SkillConfig(name="dummy", author="dummy")
    skill = Skill(configuration=config)
    skill_context = SkillContext(skill=skill)
    # dummy agent context to utilize shared state
    agent_context = AgentContext(
        identity=None,
        connection_status=None,
        outbox=None, 
        decision_maker_message_queue=None,
        decision_maker_handler_context=None,
        task_manager=None,
        default_ledger_id=None,
        currency_denominations=None,
        default_connection=None,
        default_routing=None,
        search_service_address=None,
        decision_maker_address=None,
        data_dir=None,
    )
    skill_context.set_agent_context(agent_context)
    skill_context.shared_state.update(memory)
    fsm = MyFSMBehaviour(name="babyAGI-loop", skill_context=skill_context)
    print("\033[95m\033[1m" + "\n***** INITIALIZE *****" + "\033[0m\033[0m")

    for key in action_types.keys():
        if key not in transitions:
            continue
        behaviour = SimpleStateBehaviour(name=key, skill_context=skill_context)
        is_initial = key == initial
        fsm.register_state(str(behaviour.name), behaviour, initial=is_initial)
        for event, target_behaviour_name in transitions[key].items():
            fsm.register_transition(
                str(behaviour.name), target_behaviour_name, event
            )

    skill.behaviours.update({fsm.name: fsm})

    return fsm, skill


# todo: move memory into agent state's _shared_state

class BabyAGI(Agent):
    """A re-implementation of the Baby AGI using the Open AEA framework."""

    def __init__(
        self,
        identity: Identity,
        memory: dict,
        connections: List[Connection] = None,
    ):
        """Initialise the agent."""
        super().__init__(identity, connections)
        fsm, _ = build_fsm_and_skill(memory)
        self.fsm = fsm

    def act(self):
        """Act implementation."""
        if self.fsm.is_done():
            print("done!")
            return
        self.fsm.act()

    def setup(self):
        # dummy setup method
        pass

    def teardown(self):
        # dummy teardown method
        pass


def run(first_task: str, objective: str):
    """Run babyAGI."""

    memory = create_memory(first_task, objective)

    # Create an identity for the agent
    identity = Identity(
        name="baby_agi", address="my_address", public_key="my_public_key"
    )

    print("\033[89m\033[1m" + "\n======== AEA-GPT ONLINE ========" + "\033[0m\033[0m")
    # Create our Agent (without connections)
    my_agent = BabyAGI(identity, memory)

    # Set the agent running in a different thread
    try:
        my_agent.start()
    except KeyboardInterrupt:
        # Shut down the agent
        my_agent.stop()
        print("\033[89m\033[1m" + "\n======== EXIT ========" + "\033[0m\033[0m")


if __name__ == "__main__":
    _, first_task, objective = sys.argv
    try:
        run(first_task, objective)
    except KeyboardInterrupt:
        pass


actions.py
"""
Actions: contains all the functions used to compose the low-high level actions for any agent programs that use these functions.
"""

import os
import openai
from typing import List, Tuple
from collections import deque
import pinecone

# pincone setup
USE_PINECONE = False # flag to set pincone usage on or off (default: False)
PINECONE_API_KEY = os.getenv("PINECONE_API_KEY")
PINECONE_ENVIRONMENT = os.getenv("PINECONE_ENVIRONMENT")
PINECONE_TABLE = os.getenv("PINECONE_TABLE")

# Create Pinecone index
DIMENSION = 1536
METRIC = "cosine"
POD_TYPE = "p1"
if USE_PINECONE and PINECONE_TABLE not in pinecone.list_indexes():
    pinecone.create_index(
        PINECONE_TABLE, dimension=DIMENSION, metric=METRIC, pod_type=POD_TYPE
    )

# Init Pinecone
pinecone.init(api_key=PINECONE_API_KEY, environment=PINECONE_ENVIRONMENT)


task_creation_template = """
You are a task creation AI that uses the result of an execution agent to create new tasks with the following objective: {objective}. 
The last completed task has the result: {result}. This result was based on this task description: {task_description}. 
These are incomplete tasks: {incomplete_tasks}. Based on the result, create new tasks to be completed by the AI system that do 
not overlap with incomplete tasks. Return the result as a list that is not numbered and simply lists each task in its own line, like:

First task
Second task
Third task

Do not include anything else except the list in your response.
"""
task_prioritization_template = """
You are a task prioritization AI tasked with cleaning the formatting of and reprioritizing the following tasks: {task_names}. Consider the ultimate objective of your team: {objective}. 
Do not remove any tasks. Return the result as a numbered list, like:

#. First task
#. Second task

Start the task list with number 1 and do not include anything else except the list in your response.
"""
task_execution_template = """
You are an AI that performs one task based on the following objective: {objective}.\nTake into account these previously completed tasks: {context}...and your assigned task: {task}\n. 
What is your response? Make sure to respond with a detailed solution to the assigned task you have been given only, and do not address any other tasks or make any lists. 
your response should be in paragraph form.
"""
task_stop_or_not_template = """
You are an AI that assess task completion for the following objective: {objective}. Take into account these previously completed tasks: {context}.
Has the objective been achieved? Answer with only yes or no. Only answer with yes if you think this is the best answer possible.
"""


def task_execution_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the execution prompt for GPT to take in as input when executing a task.
    It also prints the next task in the task list.

    Args:
        globals_ (dict): The globals dictionary
    
    Returns:
        str: The prompt for GPT task execution
    """
    task_list = globals_["task_list"]
    task = task_list.popleft()
    globals_["current_task"] = task

    print("\033[92m\033[1m" + "\n***** NEXT TASK *****\n" + "\033[0m\033[0m")
    print(str(task["id"]) + ": " + task["name"])

    context = get_context(globals_)
    return task_execution_template.format(
        objective=globals_["objective"], task=globals_["current_task"], context=context
    )


def task_execution_handler(response: str, globals_: dict) -> None:
    """
    This function handles the GPT response corresponding to the last task execution, allows for the
    result to be further enriched and prints the resultant GPT response from executing the task. It also prints the result

    Args:
        response (str): The GPT response from task execution
        globals_ (dict): The globals dictionary
    """
    enriched_result = {
        "data": response
    }  # This is where you should enrich the result if needed
    globals_["result"] = enriched_result

    """Use Pinecone, not currently setup"""
    id_ = globals_["current_task"]["id"]
    result_id = f"result_{id_}"
    vector = enriched_result["data"]  # extract the actual result from the dictionary
    if USE_PINECONE:
        index = pinecone.Index(index_name=PINECONE_TABLE)
        index.upsert(
            [
                (
                    result_id,
                    get_ada_embedding(vector),
                    {"task": globals_["current_task"]["name"], "result": response},
                )
            ]
        )

    print("\033[93m\033[1m" + "\n***** TASK RESULT *****\n" + "\033[0m\033[0m")
    print(globals_["result"]["data"])

    return "done"


def task_creation_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the prompt for GPT task creation so GPT can create task lists.

    Args:
        globals_ (dict): The globals dictionary

    Returns:
        str: The prompt for GPT task creation
    """
    incomplete_tasks = [t["name"] for t in globals_["task_list"]]
    return task_creation_template.format(
        objective=globals_["objective"],
        result=globals_["result"],
        task_description=globals_["current_task"].get("name", "default"),
        incomplete_tasks=incomplete_tasks,
    )


def task_creation_handler(response: str, globals_: dict):
    """
    This function handles the GPT response corresponding to the task creation prompt built by the task creation prompt builder and
    prints the resultant GPT response that is creating new tasks.

    Args:
        response (str): The GPT response from task creation
        globals_ (dict): The globals dictionary

    Returns:
        str: The status of the task creation handler

    """
    new_tasks = response.split("\n")
    if len(globals_["task_list"]) > 0:
        id_ = globals_["task_list"][-1]["id"] + 1
    else:
        id_ = 1
    task_list = [{"id": id_ + i, "name": task_name} for i, task_name in enumerate(new_tasks)]
    globals_["task_list"] = deque(task_list)

    print("\033[89m\033[1m" + "\nTASK LIST:" + "\033[0m\033[0m")
    for t in task_list:
        print(t["name"])
    
    return "done"


def task_prioritization_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the prompt for GPT task prioritization so existing tasks can be re-prioritized.

    Args:
        globals_ (dict): The globals dictionary

    Returns:
        str: The prompt for GPT task prioritization
    """
    task_list = globals_["task_list"]
    current_task = globals_["current_task"]
    task_names = [t["name"] for t in task_list]
    current_task_id = int(current_task["id"]) + 1
    objective = globals_["objective"]
    return task_prioritization_template.format(
        task_names=task_names, objective=objective, starting_id=current_task_id
    )


def task_prioritization_handler(response: str, globals_: dict):
    """
    This function handles the GPT response corresponding to the task prioritization prompt built by the task prioritization prompt builder and
    prints the resultant GPT response that is re-prioritizing existing tasks.
    """
    new_tasks = response.split("\n")
    task_list = deque([])
    for task_string in new_tasks:
        task_parts = task_string.strip().split(".", 1)
        if len(task_parts) == 2:
            task_id = int(task_parts[0].strip())
            task_name = task_parts[1].strip()
            task_list.append({"id": task_id, "name": task_name})
    globals_["task_list"] = task_list
    globals_["current_task"] = {}
    print("\033[94m\033[1m" + "\n***** RE-PRIORITIZED TASK LIST *****\n" + "\033[0m\033[0m")
    for t in task_list:
        print(str(t["id"]) + ": " + t["name"])
    return "done"


def get_ada_embedding(text):
    text = text.replace("\n", " ")
    return openai.Embedding.create(input=[text], model="text-embedding-ada-002")[
        "data"
    ][0]["embedding"]


def get_context(globals_: dict) -> List[Tuple[str]]:
    """
    Get the current context (task list) from the dictionary state variable, globals_
    """
    """Use Pinecone, not currently setup"""
    if USE_PINECONE:
        query = globals_["objective"]
        query_embedding = get_ada_embedding(query)
        index = pinecone.Index(index_name=PINECONE_TABLE)
        results = index.query(query_embedding, top_k=5, include_metadata=True)
        sorted_results = sorted(results.matches, key=lambda x: x.score, reverse=True)
        return [(str(item.metadata["task"])) for item in sorted_results]
    return globals_["task_list"]


def task_stop_or_not_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the task stop or not prompt for GPT in order to reason about the objective completeness when 
    the user stops the agent loop.

    Args:
        globals_ (dict): The globals dictionary

    Returns:
        str: The prompt for GPT task stop or not
    """
    context = get_context(globals_)
    return task_stop_or_not_template.format(
        objective=globals_["objective"], context=context
    )


def task_stop_or_not_handler(response: str, globals_: dict) -> None:
    """
    This function handles the GPT response corresponding to the task stop or not prompt built by the task stop or not prompt builder and
    prints the resultant GPT response that is reasoning about the objective completeness when the user stops the agent loop.
    """
    globals_["keep_going"] = response.strip().lower() != "yes"
    print(
        "\033[94m\033[1m" + "\n*****TASK CONTINUATION*****\n" + "\033[0m\033[0m"
    )
    print(globals_["keep_going"])
    return "done" if globals_["keep_going"] else "stop"

aea_gpt.py
import os
import sys
import time
from threading import Thread

from aea_ledger_ethereum import EthereumCrypto

from aea.aea_builder import AEABuilder
from aea.configurations.base import SkillConfig
from aea.crypto.helpers import PRIVATE_KEY_PATH_SCHEMA, create_private_key
from aea.helpers.file_io import write_with_lock
from aea.skills.base import Skill

from agent import build_fsm_and_skill, create_memory

ROOT_DIR = "./dev/autonolas/valory/langchain-experiments/task/"
INPUT_FILE = "input_file"
OUTPUT_FILE = "output_file"
PRIVATE_KEY_FILE = PRIVATE_KEY_PATH_SCHEMA.format(EthereumCrypto.identifier)

# Create a private key
create_private_key(EthereumCrypto.identifier, PRIVATE_KEY_FILE)

def build_aea(first_task: str, objective: str):

    # instantiate the aea builder
    builder = AEABuilder()
    # set the aea name
    builder.set_name("baby_agi")
    # create the shared state object that serves as memory for the actions of the AEA
    memory = create_memory(first_task, objective)
    # add the AEA's private key
    builder.add_private_key(EthereumCrypto.identifier, PRIVATE_KEY_FILE)
    # add the babyagi skill
    _, skill = build_fsm_and_skill(memory)
    # add the skill to the AEA
    builder.add_component_instance(skill)

    print("\033[89m\033[1m" + "\n======== AEA-GPT ONLINE ========" + "\033[0m\033[0m")

    # Create our AEA
    my_aea = builder.build()
    # Set the AEA's agent context
    skill.skill_context.set_agent_context(my_aea.context)
    # update the shared state of the AEA with the memory object we created above
    skill.skill_context.shared_state.update(memory)
    return my_aea


def run(first_task: str, objective: str):
    """Run babyAGI."""

    # Ensure the input and output files do not exist initially
    if os.path.isfile(INPUT_FILE):
        os.remove(INPUT_FILE)
    if os.path.isfile(OUTPUT_FILE):
        os.remove(OUTPUT_FILE)

    # Create our AEA
    my_aea = build_aea(first_task, objective)

    # Set the AEA running in a different thread
    try:
        t = Thread(target=my_aea.start)
        t.start()
    except KeyboardInterrupt:
        # Shut down the AEA
        my_aea.stop()
        t.join()
        t = None
        print("\033[89m\033[1m" + "\n======== EXIT ========" + "\033[0m\033[0m")


if __name__ == "__main__":
    _, first_task, objective = sys.argv
    try:
        run(first_task, objective)
    except KeyboardInterrupt:
        pass

simple_agent.py
"""
Heavily inspired by: https://github.com/yoheinakajima/babyagi

python simple_agent.py "develop a task list" "solve world hunger"
"""

import os
import sys
import openai
# import pinecone
from typing import List, Tuple, Any
from collections import deque
from dotenv import load_dotenv
import time

load_dotenv()

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")

# Set up OpenAI API key
openai.api_key = OPENAI_API_KEY

STOP_PROCEDURE = False

"""Use Pinecone, not currently setup"""
# PINECONE_API_KEY = os.getenv("PINECONE_API_KEY")
# PINECONE_ENVIRONMENT = os.getenv("PINECONE_ENVIRONMENT")
# PINECONE_TABLE = os.getenv("PINECONE_TABLE")

# USE_PINECONE = False

# # Init Pinecone
# pinecone.init(api_key=PINECONE_API_KEY, environment=PINECONE_ENVIRONMENT)

# # Create Pinecone index
# DIMENSION = 1536
# METRIC = "cosine"
# POD_TYPE = "p1"
# if USE_PINECONE and PINECONE_TABLE not in pinecone.list_indexes():
#     pinecone.create_index(
#         PINECONE_TABLE, dimension=DIMENSION, metric=METRIC, pod_type=POD_TYPE
#     )


task_creation_template = """
You are a task creation AI that uses the result of an execution agent to create new tasks with the following objective: {objective}. 
The last completed task has the result: {result}. This result was based on this task description: {task_description}. 
These are incomplete tasks: {incomplete_tasks}. Based on the result, create new tasks to be completed by the AI system that do 
not overlap with incomplete tasks. Return the result as a list that is not numbered and simply lists each task in its own line, like:

First task
Second task
Third task

Do not include anything else except the list in your response.
"""
task_prioritization_template = """
You are a task prioritization AI tasked with cleaning the formatting of and reprioritizing the following tasks: {task_names}. Consider the ultimate objective of your team: {objective}. 
Do not remove any tasks. Return the result as a numbered list, like:

#. First task
#. Second task

Start the task list with number 1 and do not include anything else except the list in your response.
"""
task_execution_template = """
You are an AI that performs one task based on the following objective: {objective}.\nTake into account these previously completed tasks: {context}...and your assigned task: {task}\n. 
What is your response? Make sure to respond with a detailed solution to the assigned task you have been given only, and do not address any other tasks or make any lists. 
your response should be in paragraph form.
"""
task_stop_or_not_template = """
You are an AI that assess task completion for the following objective: {objective}. Take into account these previously completed tasks: {context}.
Has the objective been achieved? Answer with only yes or no. Only answer with yes if you think this is the best answer possible.
"""


def task_execution_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the execution prompt for GPT to take in as input when executing a task.
    It also prints the next task in the task list.

    Args:
        globals_ (dict): The globals dictionary
    
    Returns:
        str: The prompt for GPT task execution
    """
    task_list = globals_["task_list"]
    task = task_list.popleft()
    globals_["current_task"] = task

    print("\033[92m\033[1m" + "\n***** NEXT TASK *****\n" + "\033[0m\033[0m")
    print(str(task["id"]) + ": " + task["name"])

    context = get_context(globals_)
    return task_execution_template.format(
        objective=globals_["objective"], task=globals_["current_task"], context=context
    )


def task_execution_handler(response: str, globals_: dict) -> None:
    """
    This function handles the GPT response corresponding to the last task execution, allows for the
    result to be further enriched and prints the resultant GPT response from executing the task. It also prints the result

    Args:
        response (str): The GPT response from task execution
        globals_ (dict): The globals dictionary
    """
    enriched_result = {
        "data": response
    }  # This is where you should enrich the result if needed
    globals_["result"] = enriched_result

    """Use Pinecone, not currently setup"""
    #id_ = globals_["current_task"]["id"]
    #result_id = f"result_{id_}"
    #vector = enriched_result["data"]  # extract the actual result from the dictionary
    #if USE_PINECONE:
    #    index = pinecone.Index(index_name=PINECONE_TABLE)
    #    index.upsert(
    #        [
    #            (
    #                result_id,
    #                get_ada_embedding(vector),
    #                {"task": globals_["current_task"]["name"], "result": response},
    #            )
    #        ]
    #    )

    print("\033[93m\033[1m" + "\n***** TASK RESULT *****\n" + "\033[0m\033[0m")
    print(globals_["result"]["data"])

    return "done"


def task_creation_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the prompt for GPT task creation so GPT can create task lists.

    Args:
        globals_ (dict): The globals dictionary

    Returns:
        str: The prompt for GPT task creation
    """
    incomplete_tasks = [t["name"] for t in globals_["task_list"]]
    return task_creation_template.format(
        objective=globals_["objective"],
        result=globals_["result"],
        task_description=globals_["current_task"].get("name", "default"),
        incomplete_tasks=incomplete_tasks,
    )


def task_creation_handler(response: str, globals_: dict):
    """
    This function handles the GPT response corresponding to the task creation prompt built by the task creation prompt builder and
    prints the resultant GPT response that is creating new tasks.

    Args:
        response (str): The GPT response from task creation
        globals_ (dict): The globals dictionary

    Returns:
        str: The status of the task creation handler

    """
    new_tasks = response.split("\n")
    if len(globals_["task_list"]) > 0:
        id_ = globals_["task_list"][-1]["id"] + 1
    else:
        id_ = 1
    task_list = [{"id": id_ + i, "name": task_name} for i, task_name in enumerate(new_tasks)]
    globals_["task_list"] = deque(task_list)

    print("\033[89m\033[1m" + "\nTASK LIST:" + "\033[0m\033[0m")
    for t in task_list:
        print(t["name"])
    
    return "done"


def task_prioritization_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the prompt for GPT task prioritization so existing tasks can be re-prioritized.

    Args:
        globals_ (dict): The globals dictionary

    Returns:
        str: The prompt for GPT task prioritization
    """
    task_list = globals_["task_list"]
    current_task = globals_["current_task"]
    task_names = [t["name"] for t in task_list]
    current_task_id = int(current_task["id"]) + 1
    objective = globals_["objective"]
    return task_prioritization_template.format(
        task_names=task_names, objective=objective, starting_id=current_task_id
    )


def task_prioritization_handler(response: str, globals_: dict):
    """
    This function handles the GPT response corresponding to the task prioritization prompt built by the task prioritization prompt builder and
    prints the resultant GPT response that is re-prioritizing existing tasks.
    """
    new_tasks = response.split("\n")
    task_list = deque([])
    for task_string in new_tasks:
        task_parts = task_string.strip().split(".", 1)
        if len(task_parts) == 2:
            task_id = int(task_parts[0].strip())
            task_name = task_parts[1].strip()
            task_list.append({"id": task_id, "name": task_name})
    globals_["task_list"] = task_list
    globals_["current_task"] = {}
    print("\033[94m\033[1m" + "\n***** RE-PRIORITIZED TASK LIST *****\n" + "\033[0m\033[0m")
    for t in task_list:
        print(str(t["id"]) + ": " + t["name"])
    return "done"


def get_ada_embedding(text):
    text = text.replace("\n", " ")
    return openai.Embedding.create(input=[text], model="text-embedding-ada-002")[
        "data"
    ][0]["embedding"]


def get_context(globals_: dict) -> List[Tuple[str]]:
    """
    Get the current context (task list) from the dictionary state variable, globals_
    """
    """Use Pinecone, not currently setup"""
    #if USE_PINECONE:
    #    query = objective
    #    query_embedding = get_ada_embedding(query)
    #    index = pinecone.Index(index_name=PINECONE_TABLE)
    #    results = index.query(query_embedding, top_k=5, include_metadata=True)
    #    sorted_results = sorted(results.matches, key=lambda x: x.score, reverse=True)
    #    return [(str(item.metadata["task"])) for item in sorted_results]
    return globals_["task_list"]


def task_stop_or_not_prompt_builder(globals_: dict) -> str:
    """
    This function builds and returns the task stop or not prompt for GPT in order to reason about the objective completeness when 
    the user stops the agent loop.

    Args:
        globals_ (dict): The globals dictionary

    Returns:
        str: The prompt for GPT task stop or not
    """
    context = get_context(globals_)
    return task_stop_or_not_template.format(
        objective=globals_["objective"], context=context
    )


def task_stop_or_not_handler(response: str, globals_: dict) -> None:
    """
    This function handles the GPT response corresponding to the task stop or not prompt built by the task stop or not prompt builder and
    prints the resultant GPT response that is reasoning about the objective completeness when the user stops the agent loop.
    """
    globals_["keep_going"] = response.strip().lower() != "yes"
    print(
        "\033[94m\033[1m" + "\n*****TASK CONTINUATION*****\n" + "\033[0m\033[0m"
    )
    print(globals_["keep_going"])
    return "done" if globals_["keep_going"] else "stop"


# Definition of the action types for the simple agent
action_types = {
    "task_creation": {
        "prompt_builder": task_creation_prompt_builder,
        "handler": task_creation_handler,
    },
    "task_prioritization": {
        "prompt_builder": task_prioritization_prompt_builder,
        "handler": task_prioritization_handler,
    },
    "task_execution": {
        "prompt_builder": task_execution_prompt_builder,
        "handler": task_execution_handler,
    },
    "task_stop_or_not": {
        "prompt_builder": task_stop_or_not_prompt_builder,
        "handler": task_stop_or_not_handler,
    },
}


def executor(globals_: dict, agent_type: str) -> None:
    """
    execute an action using simple agent

    Args:
        globals_ (dict): The globals dictionary
        agent_type (str): The action type to execute
    """
    # load the action type into "agent"
    agent = action_types[agent_type]
    # build the prompt for the corresponding action type
    builder_ = agent["prompt_builder"]
    # create the corresponding prompt for GPT to execute the action type "agent" and load it into "prompt"
    prompt = builder_(globals_)
    # call GPT with the corresponding "prompt" to execute the action and load the response from the "prompt" into "response"
    response = openai_call(prompt)
    # handle the response from GPT for the corresponding action type "agent"
    handler_ = agent["handler"]
    handler_(response, globals_)


def main(first_task: str, objective: str):
    # initialize the globals dictionary
    # this is simple_agent's state variable which is used to keep track of the task list, current task, and the objective so GPT can reason about them
    globals_ = {
        "objective": objective,
        "task_list": deque([]),
        "current_task": {},
        "result": {"data": ""},
        "keep_going": True,
    }
    # initialize the task list with the first task
    globals_["task_list"].append({"id": 1, "name": first_task})
    # simple agent loop
    while globals_["keep_going"]:
        # execution
        executor(globals_, "task_execution")
        # creation
        executor(globals_, "task_creation")
        # re-prioritization
        executor(globals_, "task_prioritization")
        if STOP_PROCEDURE:
            executor(globals_, "task_stop_or_not")
        time.sleep(1)


def openai_call(
    prompt: str, use_gpt4: bool = False, temperature: float = 0.5, max_tokens: int = 200
):
    if not use_gpt4:
        response = openai.Completion.create(
            engine="text-davinci-003",
            prompt=prompt,
            temperature=temperature,
            max_tokens=max_tokens,
            top_p=1,
            frequency_penalty=0,
            presence_penalty=0,
        )
        return response.choices[0].text.strip()
    else:
        messages = [{"role": "user", "content": prompt}]
        response = openai.ChatCompletion.create(
            model="gpt-4",
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            n=1,
            stop=None,
        )
        return response.choices[0].message.content.strip()


if __name__ == "__main__":
    _, first_task, objective = sys.argv
    try:
        main(first_task, objective)
    except KeyboardInterrupt:
        pass
